proposals/2025-03-25_otel-delta-temporality-support.md

```diff
@@ -0,0 +1,388 @@
+
+* Allow OTEL delta metrics to be ingested via the OTLP endpoint and stored directly.
+* Support for all OTEL metric types that can have delta temporality (sums, histograms, exponential histograms).
+* Allow delta metrics to be distinguished from cumulative metrics.
+* Allow the query engine to flag warnings when a cumulative function is used on deltas.
+
+### Audience
+
+This document is for Prometheus server maintainers, PromQL maintainers, and Prometheus users interested in delta ingestion.
+
+## Non-Goals
+
+* Support for ingesting delta metrics via other, non-OTLP means (e.g. replacing push gateway).
+* Advanced querying support for deltas (e.g. function overloading for rate()). Given that delta support is new and advanced querying also depends on other experimental features, the best approach - or whether we should even extend querying in any way - is currently unclear. However, this document does explore some possible options.
+
+These may come in later iterations of delta support.
+
+## How
+
+### Ingesting deltas
+
+When an OTLP sample has its aggregation temporality set to delta, write its value at `TimeUnixNano`.
+
+For the initial implementation, ignore `StartTimeUnixNano`. To ensure compatibility with the OTEL spec, this case should ideally be supported. A way to preserve `StartTimeUnixNano` is described in the potential future extension, [CT-per-sample](#ct-per-sample).
+
+### Chunks
+
+For the initial implementation, reuse existing chunk encodings.
+
+Delta counters will use the standard XOR chunks for float samples.
+
+Delta histograms will use native histogram chunks with the `GaugeType` counter reset hint/header. Currently the counter reset behaviour for cumulative native histograms is to cut a new chunk if a counter reset is detected. A (bucket or total) count drop is detected as a counter reset. As delta samples don’t build on top of each other, there could be many false counter resets detected and cause unnecessary chunks to be cut. Additionally counter histogram chunks have the invariant that no count ever goes down baked into their implementation.  `GaugeType` allows counts to go up and down, and does not cut new chunks on counter resets.
+
+### Delta metric type
+
+It is useful to be able to distinguish between delta and cumulative metrics. This would allow users to understand out what the raw data represents and what functions are appropiate to use. Additionally, this could allow the query engine or UIs displaying Prometheus data apply different behaviour depending on the metric type to provide meaningful output. 
+
+There are two options we are considering for how to type deltas. We propose to add both of these options as feature flags for ingesting deltas.
+
+1. `--enable-feature=otlp-delta-as-gauge-ingestion`: Ingests OTLP deltas as gauges.
+
+2. `--enable-feature=otlp-native-delta-ingestion`: Ingests OTLP deltas as a new delta "type", using a new `__temporality__` label to explicitly mark metrics as delta.
+
```

Comment:
Should we consider a more genetic label to keep track of the original OTEL type (ex: `__otel_type__`) - and so we can this information to make other decisions in the future - similar to what we are proposing on the rate function here?